[PHASE 1] Opaque integer resource accounting.

ConnorDoyle opened the pull request:
## [PHASE 1] Opaque integer resource accounting.

This change provides a simple way to advertise some amount of arbitrary countable resource for a node in a Kubernetes cluster. Users can consume these resources by including them in pod specs, and the scheduler takes them into account when placing pods on nodes. See the example at the bottom of the PR description for more info.

Summary of changes:
- Defines opaque integer resources as any resource with prefix `pod.alpha.kubernetes.io/opaque-int-resource-`.
- Prevent kubelet from overwriting capacity.
- Handle opaque resources in scheduler.
- Validate integer-ness of opaque int quantities in API server.
- Tests for above.

Feature issue: https://github.com/kubernetes/features/issues/76

Design: http://goo.gl/IoKYP1

Issues:

kubernetes/kubernetes#28312
kubernetes/kubernetes#19082

Related:

kubernetes/kubernetes#19080

CC @davidopp @timothysc @balajismaniam 

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
Added support for accounting opaque integer resources.

Allows cluster operators to advertise new node-level resources that would be
otherwise unknown to Kubernetes. Users can consume these resources in pod
specs just like CPU and memory. The scheduler takes care of the resource
accounting so that no more than the available amount is simultaneously
allocated to pods.
```
## Usage example

``` sh
$ echo '[{"op": "add", "path": "/status/capacity/pod.alpha.kubernetes.io~1opaque-int-resource-bananas", "value": "555"}]' | \
> http PATCH http://localhost:8080/api/v1/nodes/localhost.localdomain/status \
> Content-Type:application/json-patch+json
```

``` http
HTTP/1.1 200 OK
Content-Type: application/json
Date: Thu, 11 Aug 2016 16:44:55 GMT
Transfer-Encoding: chunked

{
    "apiVersion": "v1",
    "kind": "Node",
    "metadata": {
        "annotations": {
            "volumes.kubernetes.io/controller-managed-attach-detach": "true"
        },
        "creationTimestamp": "2016-07-12T04:07:43Z",
        "labels": {
            "beta.kubernetes.io/arch": "amd64",
            "beta.kubernetes.io/os": "linux",
            "kubernetes.io/hostname": "localhost.localdomain"
        },
        "name": "localhost.localdomain",
        "resourceVersion": "12837",
        "selfLink": "/api/v1/nodes/localhost.localdomain/status",
        "uid": "2ee9ea1c-47e6-11e6-9fb4-525400659b2e"
    },
    "spec": {
        "externalID": "localhost.localdomain"
    },
    "status": {
        "addresses": [
            {
                "address": "10.0.2.15",
                "type": "LegacyHostIP"
            },
            {
                "address": "10.0.2.15",
                "type": "InternalIP"
            }
        ],
        "allocatable": {
            "alpha.kubernetes.io/nvidia-gpu": "0",
            "cpu": "2",
            "memory": "8175808Ki",
            "pods": "110"
        },
        "capacity": {
            "alpha.kubernetes.io/nvidia-gpu": "0",
            "pod.alpha.kubernetes.io/opaque-int-resource-bananas": "555",
            "cpu": "2",
            "memory": "8175808Ki",
            "pods": "110"
        },
        "conditions": [
            {
                "lastHeartbeatTime": "2016-08-11T16:44:47Z",
                "lastTransitionTime": "2016-07-12T04:07:43Z",
                "message": "kubelet has sufficient disk space available",
                "reason": "KubeletHasSufficientDisk",
                "status": "False",
                "type": "OutOfDisk"
            },
            {
                "lastHeartbeatTime": "2016-08-11T16:44:47Z",
                "lastTransitionTime": "2016-07-12T04:07:43Z",
                "message": "kubelet has sufficient memory available",
                "reason": "KubeletHasSufficientMemory",
                "status": "False",
                "type": "MemoryPressure"
            },
            {
                "lastHeartbeatTime": "2016-08-11T16:44:47Z",
                "lastTransitionTime": "2016-08-10T06:27:11Z",
                "message": "kubelet is posting ready status",
                "reason": "KubeletReady",
                "status": "True",
                "type": "Ready"
            },
            {
                "lastHeartbeatTime": "2016-08-11T16:44:47Z",
                "lastTransitionTime": "2016-08-10T06:27:01Z",
                "message": "kubelet has no disk pressure",
                "reason": "KubeletHasNoDiskPressure",
                "status": "False",
                "type": "DiskPressure"
            }
        ],
        "daemonEndpoints": {
            "kubeletEndpoint": {
                "Port": 10250
            }
        },
        "images": [],
        "nodeInfo": {
            "architecture": "amd64",
            "bootID": "1f7e95ca-a4c2-490e-8ca2-6621ae1eb5f0",
            "containerRuntimeVersion": "docker://1.10.3",
            "kernelVersion": "4.5.7-202.fc23.x86_64",
            "kubeProxyVersion": "v1.3.0-alpha.4.4285+7e4b86c96110d3-dirty",
            "kubeletVersion": "v1.3.0-alpha.4.4285+7e4b86c96110d3-dirty",
            "machineID": "cac4063395254bc89d06af5d05322453",
            "operatingSystem": "linux",
            "osImage": "Fedora 23 (Cloud Edition)",
            "systemUUID": "D6EE0782-5DEB-4465-B35D-E54190C5EE96"
        }
    }
}
```

After patching, the kubelet's next sync fills in allocatable:

```
$ kubectl get node localhost.localdomain -o json | jq .status.allocatable
```

``` json
{
  "alpha.kubernetes.io/nvidia-gpu": "0",
  "pod.alpha.kubernetes.io/opaque-int-resource-bananas": "555",
  "cpu": "2",
  "memory": "8175808Ki",
  "pods": "110"
}
```

Create two pods, one that needs a single banana and another that needs a truck load:

```
$ kubectl create -f chimp.yaml
$ kubectl create -f superchimp.yaml
```

Inspect the scheduler result and pod status:

```
$ kubectl describe pods chimp
Name:           chimp
Namespace:      default
Node:           localhost.localdomain/10.0.2.15
Start Time:     Thu, 11 Aug 2016 19:58:46 +0000
Labels:         <none>
Status:         Running
IP:             172.17.0.2
Controllers:    <none>
Containers:
  nginx:
    Container ID:       docker://46ff268f2f9217c59cc49f97cc4f0f085d5ac0e251f508cc08938601117c0cec
    Image:              nginx:1.10
    Image ID:           docker://sha256:82e97a2b0390a20107ab1310dea17f539ff6034438099384998fd91fc540b128
    Port:               80/TCP
    Limits:
      cpu:                                      500m
      memory:                                   64Mi
      pod.alpha.kubernetes.io/opaque-int-resource-bananas:   3
    Requests:
      cpu:                                      250m
      memory:                                   32Mi
      pod.alpha.kubernetes.io/opaque-int-resource-bananas:   1
    State:                                      Running
      Started:                                  Thu, 11 Aug 2016 19:58:51 +0000
    Ready:                                      True
    Restart Count:                              0
    Volume Mounts:                              <none>
    Environment Variables:                      <none>
Conditions:
  Type          Status
  Initialized   True 
  Ready         True 
  PodScheduled  True 
No volumes.
QoS Class:      Burstable
Events:
  FirstSeen     LastSeen        Count   From                            SubobjectPath           Type            Reason                  Message
  ---------     --------        -----   ----                            -------------           --------        ------                  -------
  9m            9m              1       {default-scheduler }                                    Normal          Scheduled               Successfully assigned chimp to localhost.localdomain
  9m            9m              2       {kubelet localhost.localdomain}                         Warning         MissingClusterDNS       kubelet does not have ClusterDNS IP configured and cannot create Pod using "ClusterFirst" policy. Falling back to DNSDefault policy.
  9m            9m              1       {kubelet localhost.localdomain} spec.containers{nginx}  Normal          Pulled                  Container image "nginx:1.10" already present on machine
  9m            9m              1       {kubelet localhost.localdomain} spec.containers{nginx}  Normal          Created                 Created container with docker id 46ff268f2f92
  9m            9m              1       {kubelet localhost.localdomain} spec.containers{nginx}  Normal          Started                 Started container with docker id 46ff268f2f92
```

```
$ kubectl describe pods superchimp
Name:           superchimp
Namespace:      default
Node:           /
Labels:         <none>
Status:         Pending
IP:
Controllers:    <none>
Containers:
  nginx:
    Image:      nginx:1.10
    Port:       80/TCP
    Requests:
      cpu:                                      250m
      memory:                                   32Mi
      pod.alpha.kubernetes.io/opaque-int-resource-bananas:   10Ki
    Volume Mounts:                              <none>
    Environment Variables:                      <none>
Conditions:
  Type          Status
  PodScheduled  False 
No volumes.
QoS Class:      Burstable
Events:
  FirstSeen     LastSeen        Count   From                    SubobjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                    -------------   --------        ------                  -------
  3m            1s              15      {default-scheduler }                    Warning         FailedScheduling        pod (superchimp) failed to fit in any node
fit failure on node (localhost.localdomain): Insufficient pod.alpha.kubernetes.io/opaque-int-resource-bananas
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31652)

<!-- Reviewable:end -->

ConnorDoyle commented:
@balajismaniam and I are both covered by the Intel corporate CLA (org emails are public in our GitHub profiles)

timothysc commented on docs/proposals/opaque-integer-resources.md:
s/(If there are specializations that hang off of the resource name, then it is not opaque!)/ 

makes more sense to define what you mean by opaque.  In general, the name opaque lends to more confusion.  We used to call these "machine limits" in condor.

right now the definition of an opaque resource is opaque.

timothysc commented on docs/proposals/opaque-integer-resources.md:
You would need to nix the isolation bit above for this to be true.   In general, I don't think you need the clause above.

timothysc commented on docs/proposals/opaque-integer-resources.md:
nix this.

timothysc commented on docs/proposals/opaque-integer-resources.md:
So I know this PR is isolated, but if we're going to have a design doc then we need to spell that out.  You may just want to nix the design doc from this PR.

timothysc commented on docs/proposals/opaque-integer-resources.md:
/cc @derekwaynecarr

timothysc commented on pkg/api/types.go:
```diff
@@ -2262,6 +2262,11 @@ const (
 
+const (
+	// Namespace prefix for opaque counted resources (alpha).
+	ResourceOpaqueIntPrefix = "opaque-int.alpha.kubernetes.io/"
```
I'm guessing this won't fly.  Let me find a better mechanism.

davidopp replied:
@timothysc what is the issue with this?

ConnorDoyle replied:
@timothysc ping (suggestions welcome, or can we drop this issue?)

ConnorDoyle replied:
Related question, have been seeing comments lately concerned with proliferation of "alpha-prefixed" things in the API and confusion it might cause. Is there an official direction for what to do here?

ConnorDoyle replied:
Bikeshed/naming: does anyone prefer the prefix `"alpha.kubernetes.io/opaque-"` to match the existing `""alpha.kubernetes.io/nvidia-gpu"` resource?

timothysc replied:
patterns are fuzzy, but the closest thing I see is AREA.alpha.kubernetes.io/FEATURE  

Where area = things like control-plane/scheduler/pod 

I debate if this is pod?  really seems like pod to me.

ConnorDoyle replied:
Searching the source code I found at least:

- pod.alpha.kubernetes.io
- pods.alpha.kubernetes.io
- scheduler.alpha.kubernetes.io
- security.alpha.kubernetes.io
- container.seccomp.security.alpha.kubernetes.io
- seccomp.security.alpha.kubernets.io
- create.snapshot.volume.alpha.kubernetes.io
- snapshot.volume.alpha.kubernetes.io
- container.apparmor.security.alpha.kubernetes.io
- metrics.alpha.kubernetes.io
- failure-domain.alpha.kubernetes.io
- service.alpha.kubernetes.io
- metadata.annotations.scheduler.alpha.kubernetes.io
- rkt.alpha.kubernetes.io

So... pod makes sense to me. Neither scheduler nor node capture the whole scope of where these values are used. Since it's more or less a free-for-all we could probably make the case for "resource" as a new prefix too (`opaque.resource.alpha.kubernetes.io`). Alternatively, we can match the existing alpha resource as mentioned above.

timothysc replied:
I think pod.alpha.kubernetes.io/opaque-int-resource follows the precedent the best.

ConnorDoyle replied:
Will update.

ConnorDoyle replied:
Updated to `pod.alpha.kubernetes.io/opaque-int-resource-`.

timothysc commented on pkg/api/validation/validation.go:
```diff
@@ -2784,6 +2788,18 @@ func ValidateNodeUpdate(node, oldNode *api.Node) field.ErrorList {
 
+// Validate that opaque integer resource quantities in a resource list are
+// indeed integers.
+func ValidateOpaqueIntQuants(rs api.ResourceList, fldPath *field.Path) field.ErrorList {
```
So I'm not sure if this would fly with the api-machinery folks mucking the existing resource lists.

ConnorDoyle replied:
We could validate elsewhere, but it seemed nice to fail as soon as possible. Is there another way we can do this?

ConnorDoyle replied:
@timothysc ping (is there someone from api-machinery who can say if this is OK to do?)

derekwaynecarr replied:
ValidateResourceQuantityValue in validation.go does something similar.  I would use that same pattern and just move integer checking into a shared call

ConnorDoyle replied:
Consolidated the validation logic a bit.

timothysc commented on pkg/api/validation/validation_test.go:
```diff
@@ -3575,6 +3576,52 @@ func TestValidatePod(t *testing.T) {
+						ImagePullPolicy: "IfNotPresent",
+						Resources: api.ResourceRequirements{
+							Requests: api.ResourceList{
+								api.ResourceName(fmt.Sprintf("%sopaque-A", api.ResourceOpaqueIntPrefix)): resource.MustParse("10"),
```
probably abstract this into a utility f(n)

ConnorDoyle replied:
Yep, good idea.

ConnorDoyle replied:
Added helper.

timothysc commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -341,23 +341,32 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
+	// Note: avoid blindly overwriting the capacity in case opaque
+	//       resources are being advertised.
+	if node.Status.Capacity == nil {
+		node.Status.Capacity = api.ResourceList{}
```
umm I think there are edge cases here perhaps on over-provisioning. 
/cc @derekwaynecarr

ConnorDoyle replied:
Can you elaborate? Previously this function unconditionally clobbered `node.Status.Capacity`.

derekwaynecarr replied:
the set call looks fine here, but I think it may be still possible that Kubelet will have node capacity cached elsewhere and will never observe the opaque resources were added.   I am not sure if that could result in failed admission checks.  I will need to vet that part of the code when not on phone

ConnorDoyle replied:
@derekwaynecarr, do you still need to do some additional vetting of the affected Kubelet code? I checked for this during initial implementation, but If it would be helpful I could do an audit/writeup.

derekwaynecarr replied:
@ConnorDoyle -- apologies for the delay, too much going on.

During kubelet admission, we call a registered set of admit handlers to do admission checks that are registered.

see: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L724

this function ends up using `getNodeAnyway`, which falls through to this function you modified eventually, so I think we are ok for the area that would actually be impacted.

timothysc commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -341,23 +341,32 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
+		node.Status.Capacity[api.ResourceCPU] = *resource.NewMilliQuantity(0, resource.DecimalSI)
+		node.Status.Capacity[api.ResourceMemory] = resource.MustParse("0Gi")
+		node.Status.Capacity[api.ResourcePods] = *resource.NewQuantity(int64(kl.maxPods), resource.DecimalSI)
+		node.Status.Capacity[api.ResourceNvidiaGPU] = *resource.NewQuantity(int64(kl.nvidiaGPUs), resource.DecimalSI)
```
I don't think we can/should add elements unless they exist.

ConnorDoyle commented on docs/proposals/opaque-integer-resources.md:
We can drop the doc part from this PR. It's mostly verbatim from the Google doc, and we've linked to it from the description. When we can resolve the outstanding questions there (and here) we'll reintroduce it.

ConnorDoyle commented on docs/proposals/opaque-integer-resources.md:
Dropped the doc.

davidopp commented:
Please split the code and proposal into separate PRs.

davidopp commented:
Never mind, I thought I saw an .md file in here, but it's not there anymore. Weird.

davidopp commented:
My concern with this approach is that IIUC you're relying on getting the opaque resources from `cadvisor.CapacityFromMachineInfo`. One approach we talked about, which I think would possibly be better, is to have a daemon pod on the node do a patch to the node capacity. Then you can put all of the logic for detecting opaque resources and determining their capacity in your own container, and other people can do the same thing. Is there some reason why the opaque resouces have to be known to cAdvisor (other than to make the approach you're taking here work)?

ConnorDoyle commented:
Pulled it out after similar (hidden above) comments from Tim.

> On Sep 1, 2016, at 22:49, David Oppenheimer <notifications@github.com> wrote:
> 
> Never mind, I thought I saw an .md file in here, but it's not there anymore. Weird.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ConnorDoyle commented:
The change to the cAdvisor capacity code in the Kubelet just prevents the patched capacity from being overwritten. There isn't any code yet to discover the counted resources and do the patch. We are picking it up assuming the patch request comes from an external source, and probably a discovery daemon or similar method like you describe.

> On Sep 1, 2016, at 23:17, David Oppenheimer <notifications@github.com> wrote:
> 
> My concern with this approach is that IIUC you're relying on getting the opaque resources from cadvisor.CapacityFromMachineInfo. One approach we talked about, which I think would possibly be better, is to have a daemon pod on the node do a patch to the node capacity. Then you can put all of the logic for detecting opaque resources and determining their capacity in your own container, and other people can do the same thing. Is there some reason why the opaque resouces have to be known to cAdvisor (other than to make the approach you're taking here work)?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

balajismaniam commented:
@ConnorDoyle We can add the usage example (from the .md file) in the PR description. This will explain how we are advertising the resources.

ConnorDoyle commented:
@balajismaniam great suggestion. Updated. @davidopp please see the example in the updated PR description at the top of the page.

derekwaynecarr commented:
I have no cycles to review this until 1.4 settles down but I would like to
hold this PR until I can review early in 1.5.  Thanks

On Friday, September 2, 2016, Connor Doyle <notifications@github.com> wrote:

> @balajismaniam <https://github.com/balajismaniam> great suggestion.
> Updated. @davidopp <https://github.com/davidopp> please see the example
> in the updated PR description at the top of the page.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-244433403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbIfEthnd72iawFl1ZxxqDSuAC4j4ks5qmFhigaJpZM4Jv_NU>
> .
>

ingvagabund commented:
Skipping the tests the node and the scheduler parts are straightforward. Did not found any obvious trap.

timothysc commented:
ok, let start to move this forwards.  

@ConnorDoyle could you please do the following: 
1. Squash commits 
2. Address all comments 
3. Update title to outline Phase 1 of Feature

ConnorDoyle commented:
@timothysc squashed and addressed comments where there was clear resolution with more commits. Can squash again on request. There are a couple less specific concerns outstanding that I'm not sure how to resolve (I asked for clarification above). Updated the PR title.

ConnorDoyle commented:
@timothysc looks like CLA bot overwrote your label... maybe `cla: human-approved` instead? Not sure why it's not picking up the org though.

timothysc commented:
@ConnorDoyle re: CLA, you're going to need to talk with google folks on that one. 

/cc @eparis

derekwaynecarr reviewed (commented):
This looks mostly good to me, just need to check where else we are caching machine info in Kubelet

ConnorDoyle committed:
Support opaque integer resource accounting.

- Prevents kubelet from overwriting capacity during sync.
- Handles opaque integer resources in the scheduler.
  - Adds scheduler predicate tests for opaque resources.
- Validates opaque int resources:
  - Ensures supplied opaque int quantities in node capacity,
    node allocatable, pod request and pod limit are integers.
  - Adds tests for new validation logic (node update and pod spec).
- Added e2e tests for opaque integer resources.

ConnorDoyle commented:
Reset author email in commits to satisfy CLA bot.

timothysc commented:
@ConnorDoyle I think it's much closer only minor comment from me plus follow up re: capability maintaining on the node side.  

However, I do think this needs an e2e that can verify the full behavior.

ConnorDoyle commented:
@k8s-bot unit test this

ConnorDoyle commented:
@k8s-bot gke e2e test this

ConnorDoyle commented:
@k8s-bot kubemark e2e test this

ConnorDoyle commented:
@k8s-bot gci gke e2e test this

ConnorDoyle commented:
@timothysc agree we need some e2e tests for the feature. Do you mind if we address that in a follow-up PR?

ConnorDoyle commented:
@k8s-bot unit test this
@k8s-bot gci gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot gke e2e test this

thockin commented:
WRT APi we should think about how this will ultimately be represented.  I think I like something like a map[string]int or list[struct{string, int}].  That suggests to me that the annotation should not be a prefix (I don't think we have any precedent for that) but a literal string, and the value should be a JSOn-encoded map or list.

ConnorDoyle commented:
Hey @thockin, I didn't quite understand what you're suggesting. Probably I'm missing something, but just to be sure I'll try to restate what's happening here.

We're not using an annotation here to represent the new resources. The prefix in this proposed changed is compared against the name of a resource that that operators advertise as part of the node capacity, and that users request in their pod spec. The resource name is used in `Node.Status.{Capacity, Allocatable}` and also in `ResourceRequirements.{Limits, Requests}`. In either case the name ends up being a key in an `api.ResourceList` which is a type alias of `map[api.ResourceName]resource.Quantity` (where `api.ResourceName` is an alias of type `string`.

Could you describe where the map or list would live in the API?

davidopp commented:
+1 to what @ConnorDoyle said.

thockin commented:
I did mis-read, but I still don't see why it needs this prefix

On Sat, Oct 1, 2016 at 11:17 PM, David Oppenheimer <notifications@github.com
> wrote:

> +1 to what @ConnorDoyle <https://github.com/ConnorDoyle> said.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-250956078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVArYgGUuuFljYSm4qQG0jQu-k_boks5qv0x4gaJpZM4Jv_NU>
> .
>

ConnorDoyle commented:
Totally fair question. Here's our line of reasoning (not to say it's necessarily valid.) The prefix "kind of" follows the precedent set by the NvidiaGpus resource. Since the feature is alpha the resource name makes that explicit. In our case support for _any_ opaque resource is alpha. Additionally, since resources don't have explicit types and are all divisible to milli-units, the prefix restricts the quantities to integer values (in case you don't want to support requests for 0.1 of an FPGA device for example.)

Would you rather see the system treat any unrecognized resource name as an opaque resource? The integer restriction could be enforced by the operator via admission controllers in that case.

> On Oct 2, 2016, at 20:24, Tim Hockin <notifications@github.com> wrote:
> 
> I did mis-read, but I still don't see why it needs this prefix
> 
> On Sat, Oct 1, 2016 at 11:17 PM, David Oppenheimer <notifications@github.com
> > wrote:
> 
> > +1 to what @ConnorDoyle <https://github.com/ConnorDoyle> said.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-250956078>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AFVgVArYgGUuuFljYSm4qQG0jQu-k_boks5qv0x4gaJpZM4Jv_NU>
> > .
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidopp commented:
> Would you rather see the system treat any unrecognized resource name as an opaque resource?

I think this is a bad idea because then we couldn't do any meaningful validation to catch typos.

thockin commented:
I don't think (in the end) that these opaque resources should be any
different from builtins, when it comes to end-user experience.  Users
should not have to type "opaque-int-foobar" to get a foobar.

On Sun, Oct 2, 2016 at 9:42 PM, David Oppenheimer <notifications@github.com>
wrote:

> Would you rather see the system treat any unrecognized resource name as an
> opaque resource?
>
> I think this is a bad idea because then we couldn't do any meaningful
> validation to catch typos.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-251025576>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVAL_IP2Hjefe8KP5LYGkg_6TUhj4ks5qwIe0gaJpZM4Jv_NU>
> .
>

timothysc commented:
> @timothysc agree we need some e2e tests for the feature.   Do you mind if we address that in a follow-up PR?  

@ConnorDoyle That kind of flies against precedent, although we've made exceptions in the past, ideally i would not like to do that on this one, given that we are early in the cycle here and have time to vet it with automation.

timothysc commented:
> Users should not have to type "opaque-int-foobar" to get a foobar.

Totally agree in the long run... In the interm alpha-phase, I'm not opposed to the current implementation, given that it looks/acts like gpus.

thockin commented:
I am fine with that

On Mon, Oct 3, 2016 at 11:48 AM, Timothy St. Clair <notifications@github.com
> wrote:

> Users should not have to type "opaque-int-foobar" to get a foobar.
>
> Totally agree in the long run... In the interm alpha-phase, I'm not
> opposed to the current implementation, given that it looks/acts like gpus.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-251191338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVPzgI4Qy5aFpwHarwyv2JOiEb7n_ks5qwU4FgaJpZM4Jv_NU>
> .
>

ConnorDoyle commented:
@timothysc OK, I'll work on adding some e2e validation tests ASAP (can get to it later this week).

> On Oct 3, 2016, at 11:42, Timothy St. Clair <notifications@github.com> wrote:
> 
> @timothysc agree we need some e2e tests for the feature. Do you mind if we address that in a follow-up PR?
> 
> @ConnorDoyle That kind of flies against precedent, although we've made exceptions in the past, ideally i would not like to do that on this one, given that we are early in the cycle here and have time to vet it with automation.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ConnorDoyle commented:
@thockin @davidopp are either of you in favor of dropping the integer validation piece and renaming the feature as just opaque resources? Looking ahead if the prefix is dropped in GA it would be weird/limiting to enforce that constraint. At first I think I didn't understand admission controller usage well enough but that would seem more kubernetish.

> On Oct 3, 2016, at 12:17, Tim Hockin <notifications@github.com> wrote:
> 
> I am fine with that
> 
> On Mon, Oct 3, 2016 at 11:48 AM, Timothy St. Clair <notifications@github.com
> > wrote:
> 
> > Users should not have to type "opaque-int-foobar" to get a foobar.
> >
> > Totally agree in the long run... In the interm alpha-phase, I'm not
> > opposed to the current implementation, given that it looks/acts like gpus.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-251191338>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AFVgVPzgI4Qy5aFpwHarwyv2JOiEb7n_ks5qwU4FgaJpZM4Jv_NU>
> > .
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

thockin commented:
Are you suggesting allowing fractional values?  e.g. half of a GPU?  I
think that would be as weird as half a byte.  We just need to validate
against it or round or ...

On Mon, Oct 3, 2016 at 9:39 PM, Connor Doyle <notifications@github.com>
wrote:

> @thockin @davidopp are either of you in favor of dropping the integer
> validation piece and renaming the feature as just opaque resources? Looking
> ahead if the prefix is dropped in GA it would be weird/limiting to enforce
> that constraint. At first I think I didn't understand admission controller
> usage well enough but that would seem more kubernetish.
>
> > On Oct 3, 2016, at 12:17, Tim Hockin <notifications@github.com> wrote:
> >
> > I am fine with that
> >
> > On Mon, Oct 3, 2016 at 11:48 AM, Timothy St. Clair <
> notifications@github.com
> > > wrote:
> >
> > > Users should not have to type "opaque-int-foobar" to get a foobar.
> > >
> > > Totally agree in the long run... In the interm alpha-phase, I'm not
> > > opposed to the current implementation, given that it looks/acts like
> gpus.
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/kubernetes/kubernetes/pull/
> 31652#issuecomment-251191338>,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> AFVgVPzgI4Qy5aFpwHarwyv2JOiEb7n_ks5qwU4FgaJpZM4Jv_NU>
> > > .
> > >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-251294107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMz0lIh6ufvIuya11Ep9uKUSmgEQks5qwdiXgaJpZM4Jv_NU>
> .
>

davidopp commented:
I think Connor's point is that if there's nothing in the name that indicates it's an opaque integer resource, then we can't know to validate that it's an integer.

thockin commented:
Well I, for one, hate that we have anything as float in the first place.
That said, can we validate that any unknown value is integer?  Or just not
validate and let things play out?  If there are 2 GPU​s and someone asks
for 1.1 of them, they get 2.

davidopp commented:
Everything is expressed in milis, so the validation would actually be checking that the number is divisible by 1000.

ConnorDoyle commented:
> can we validate that any unknown value is integer? 

@thockin there is validation in the PR to do exactly that. To put it another way, do we need to support ANY fractional opaque resource in the future? Could well be that integers are all you need. (If a user really needs fractions they could just treat the units as micro-widgets or whatever when enforcing allocation.)

> On Oct 3, 2016, at 22:16, Tim Hockin <notifications@github.com> wrote:
> 
> Well I, for one, hate that we have anything as float in the first place.
> That said, can we validate that any unknown value is integer? Or just not
> validate and let things play out? If there are 2 GPU​s and someone asks
> for 1.1 of them, they get 2.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidopp commented:
Seriously I think this design is starting to go a bit off the rails.

We're proposing that if the user spells "memory" as "memoru" they should get back an error that they should make their request divisible by 1000?

davidopp commented:
Or would we at least be able to validate that the resource name is not a built-in resource nor a currently-registered opaque resource and thus reject it?

ConnorDoyle commented:
To my limited experience, a *short* (in GA) prefix seems least bad. e.g. user/foo.

> On Oct 3, 2016, at 22:35, David Oppenheimer <notifications@github.com> wrote:
> 
> Or would we at least be able to validate that the resource name is not a built-in resource nor a currently-registered opaque resource and thus reject it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

thockin commented:
My point is that the prefix should be owned  by the person/org defining the
resource.  Not kubernetes.io.

On Mon, Oct 3, 2016 at 10:39 PM, Connor Doyle <notifications@github.com>
wrote:

> To my limited experience, a *short* (in GA) prefix seems least bad. e.g.
> user/foo.
>
>
> > On Oct 3, 2016, at 22:35, David Oppenheimer <notifications@github.com>
> wrote:
> >
> > Or would we at least be able to validate that the resource name is not a
> built-in resource nor a currently-registered opaque resource and thus
> reject it?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-251300695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVNaVpugjAFIv39Z8-cCQSRfkHolyks5qweahgaJpZM4Jv_NU>
> .
>

ConnorDoyle commented:
I signed it!

timothysc commented:
@davidopp what's our policy now, given that annotations is semi-defunct?

ConnorDoyle commented:
@timothysc this change does not use annotations, does it need to be discussed here?

timothysc commented:
@ConnorDoyle Ohh whoops.. no it doesn't.

ConnorDoyle commented:
@timothysc question regarding the approach to e2e's for alpha features. I followed suit by including a `[Feature: OpaqueResources]` suffix, but apparently those get skipped in the default config executed by the PR gate. Should I remove the suffix so it runs? Kind of confused about what is expected here.

timothysc commented:
> Should I remove the suffix so it runs?

No, that's legit and there are longer run tests that will hit it.

ConnorDoyle commented:
OK, that's cool. I am going to temporarily remove the suffix just to we can see if it passes in the centralized env and will add it back once we see the output.

ConnorDoyle commented:
@k8s-bot cvm gke e2e test this

ConnorDoyle commented:
@timothysc e2e tests are passing! I will re-add the feature suffix and rebase as soon as GKE smoke turns green.

I'm authorized for the CNCF CLA via corporate signature (see screenshot) but the label isn't being applied properly.

![screen shot 2016-10-14 at 2 59 38 pm](https://cloud.githubusercontent.com/assets/379372/19406450/b3e06952-923a-11e6-84e3-282e707d519c.png)

What else is required move this forward toward merge?

ConnorDoyle commented:
Squashed into two commits (one for feature and another for e2e test).

ConnorDoyle commented:
Just for posterity, I'd like to describe a bug @balajismaniam and I uncovered with the original implementation of the feature while developing e2e tests.

At first the e2e test for this feature failed by wrongly scheduling pods that consumed too many opaque integer resources.

The problem was caused by an early return in the `PodFitsResources` scheduler predicate:

https://github.com/kubernetes/kubernetes/blob/182e89b84edf407ebbe56f0cffa117f6a47c8f36/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L465

Our manual smoke testing consisted of creating pods in the default namespace. The LimitRanger in its default configuration ensured that these pods had a minimum amount of CPU resource, which caused the guard on this early return to fail.

However, the e2e tests run pods in a namespace per test case so pods with only opaque resource requests hit the early return and were admitted. The feature commit in this PR now amends the guard to look for opaque integer resources.

I'm glad for the experience since we learned a lot about the e2e framework, but it was pretty painful and slow to debug the test. Many thanks to @balajismaniam for the pair-debugging session.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,309 @@
+			// TODO(CD): Watch scheduler events instead to catch an indication of
+			//           insufficient resources or other unschedulable condition
+			//           instead of waiting for this to time out.
+			Expect(pod).To(notBeScheduled())
```
There are other tests that do this per-similar request: 

see: https://github.com/kubernetes/kubernetes/pull/26501/files#diff-331b6cbc690d4629a538fe6b8d293bc1R69

ConnorDoyle replied:
Sorry, the TODO is out-dated. I asked around in sig-testing for recommendations and the only feedback I got was not to rely on events. This doesn't wait for the `beScheduled` predicate to time out, rather it looks for the unschedulable condition on the pod.

I'm going to delete the comment here because it's not relevant. If you would still like these to be refactored in light of the above please let me know.

ConnorDoyle replied:
Not to push too hard in the appeal, but the check implemented here gives _great_ error messages on failure.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,309 @@
+			}
+			pod, err = f.Client.Pods(f.Namespace.Name).Create(pod)
+			Expect(err).NotTo(HaveOccurred())
+			// TODO(CD): Watch scheduler events instead to catch an indication of
```
Same comment here.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,309 @@
+func refresh(f *framework.Framework, old *api.Node) *api.Node {
+	if old == nil {
+		// Priming invocation; select the first non-master node.
+		nodes, err := f.Client.Nodes().List(api.ListOptions{})
```
I think you can probably refactor to use the informer for all list/watch notify checks.

timothysc reviewed (commented):
Refactor e2es to use informer then lgtm to me.  

I'd like to hand this off after that to @davidopp and @derekwaynecarr for final (lgtm)

ConnorDoyle commented:
@k8s-bot gci gke e2e test this

ConnorDoyle commented:
@k8s-bot gci gke e2e test this

timothysc commented:
@davidopp @derekwaynecarr I realize folks are busy, but I'm going to timeout by next Friday on this one, because I have other things to get done for this release.

davidopp commented:
I skimmed it and I'm fine with it. Please be sure to write user documentation (or more correctly, admin documentation) before the code deadline.

derekwaynecarr commented:
@ConnorDoyle -- is there a use case today for an opaque resource that has a request and limit value that is not equal?  If so, can you elaborate on it?  If we have support for opaque resources, I want to have a plan for how I can control their consumption via `ResourceQuota` and I did not have a clear case in mind that would have the request/limit value different.

ConnorDoyle commented:
@derekwaynecarr, so far none of the use cases I know of would need request ≠ limit, but IMO it's not something we should rule out.

> If so, I would like to reserve the pod.alpha.kubernetes.io/requests. and pod.alpha.kubernetes.io/limits. prefixes from being used as an opaque resource name.

This is already the case, since the prefix is the whole string `pod.alpha.kubernetes.io/opaque-int-resource-`. So `pod.alpha.kubernetes.io/{limits|requests}...` can't be an opaque int resource name.

derekwaynecarr commented:
@ConnorDoyle - my bad, I should have remembered that.  Blame context
switching!

On Wednesday, October 19, 2016, Connor Doyle <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr>, so far none of the
> use cases I know of would need request ≠ limit, but IMO it's not something
> we should rule out.
>
> If so, I would like to reserve the pod.alpha.kubernetes.io/requests. and
> pod.alpha.kubernetes.io/limits. prefixes from being used as an opaque
> resource name.
>
> This is already the case, since the prefix is the whole string
> pod.alpha.kubernetes.io/opaque-int-resource-. So pod.alpha.kubernetes.io/{
> limits|requests}.
> <http://pod.alpha.kubernetes.io/%7Blimits%7Crequests%7D.>.. can't be an
> opaque int resource name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31652#issuecomment-254892410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbB51Q7qg2iY0NBIKLfFiCD9h_ZMwks5q1lsmgaJpZM4Jv_NU>
> .
>

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

ConnorDoyle commented:
@k8s-bot verify test this

ConnorDoyle commented:
@timothysc PTAL. Refactored tests using informer turned green in all e2e environments, re-added the `[Feature:...` suffix to the description.

Not sure what's up with the verification check.  In CI there seems to be a problem with `hack/verify_bazel.sh`.

ConnorDoyle commented:
Fixed up the verification check (thanks to @mikedanese.)

timothysc commented:
@ConnorDoyle please squash commits.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,373 @@
+		Expect(err).NotTo(HaveOccurred())
+		Expect(success).To(Equal(true))
+
+		By("Creating a pod with two containers that together require more of the opaque resource than is allotable on any node")
```
allocatable.

ConnorDoyle replied:
Reviewed strings w/ `aspell`, this is the only one I could find as well.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,373 @@
+	informerStartedChan := make(chan struct{})
+	var informerStartedGuard sync.Once
+
+	_, controller := cache.NewInformer(
```
Thanks, It eliminates all the polling too.

ConnorDoyle commented on test/e2e/opaque_resource.go:
```diff
@@ -0,0 +1,373 @@
+	return strings.Replace(string(resName), "/", "~1", -1)
+}
+
+// Returns true if an node update matching the predicate was emitted from the
```
nit: s/an/a

ConnorDoyle commented:
@timothysc squashed, just awaiting CI results then everything should be good to go. Thanks all for reviews.

ConnorDoyle commented:
@timothysc checks are green, could you please add your LGTM?

timothysc commented:
@ConnorDoyle could you add your release note above too please.

vishh commented:
LGTM

ConnorDoyle commented:
@timothysc added summary line and extended release note in the PR description.

ConnorDoyle committed:
Updated test owners file.

foxish commented:
@ConnorDoyle You're failing Jenkins verification. I think you need to run ./hack/update_owners.py to fix that.

ConnorDoyle commented:
@foxish thanks, just pushed a commit for that. The verification script is passing locally for me now.

RenaudWasTaken commented:
@ConnorDoyle in your example, the superchimp pod gets a "FailedScheduling" status.

However, the behavior for CPU and memory seem to be to put the pod in a "Pending" state which means that when the resource becomes available the pod can get rescheduled.

It seems to me that the current behavior for opaque resource means that pods which request more resources than available fail to be rescheduled once the resources becomes available once again is that correct ?

Is it the expected behavior and/or is there a way to get the same behavior as CPU/memory 😄?

ConnorDoyle commented:
@RenaudWasTaken the behavior should be the same as CPU and memory, if you see something else happening then it's a bug. The example in the PR description may not have been kept up to date since the very first version of the patch set.

> On Feb 17, 2017, at 16:06, RenaudWasTaken <notifications@github.com> wrote:
> 
> @ConnorDoyle in your example, the superchimp pod gets a "FailedScheduling" status.
> 
> However, the behavior for CPU and memory seem to be to put the pod in a "Pending" state which means that when the resource becomes available the pod can get rescheduled.
> 
> It seems to me that the current behavior for opaque resource means that pods which request more resources than available fail to be rescheduled once the resources becomes available once again is that correct ?
> 
> Is it the expected behavior and/or is there a way to get the same behavior as CPU/memory 😄?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
